lmp/model/_transformer.py

```diff
@@ -0,0 +1,460 @@
+r"""Transformer Language Model with Transformer's encoder architecture."""
+
+
+import math
+import argparse
+from typing import ClassVar, Dict, Optional
+
+import torch
+import torch.nn as nn
+
+from lmp.model._rnn import RNNModel
+from lmp.tknzr._base import BaseTknzr
+
+
+class PositionalEncoding(nn.Module):
+    r"""Positional Encoding.
+
+    In order for the model to make use of the order of the sequence,
+    add ``positional encodings`` to the input embeddings.
+
+    .. math::
+
+            \begin{align}
+            \text{PE}_{(\text{pos},\text{2i})} &= \sin\
+             (\frac{\text{pos}}{10000^\frac{\text{2i}}{d_{\text{model}}}}) \\
+            \text{PE}_{(\text{pos},\text{2i+1})} &= \cos\
+             (\frac{\text{pos}}{10000^\frac{\text{2i}}{d_{\text{model}}}}) \\
+            \end{align}
+
+    Parameters
+    ==========
+    d_hid: int
+        Hidden dimension for MLP and Transformer.
+        Must be bigger than or equal to ``1``.
+    dropout: float
+        Dropout probability.
+        Must satisfy ``0.0 <= dropout <= 1.0``.
+    max_len: int
+        Max length of the input sequence.
+        Must be bigger than or equal to ``1``.
+
+    Attributes
+    ==========
+    dropout: torch.nn.Dropout
+        Positional Encoding dropout.
+    pe: torch.nn.FloatTensor
+        The values of Positional Encoding.
+    """
+
+    def __init__(self, d_hid, dropout, max_len):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        # Create positional encoding table.
+        # Shape : `(S, H)`
+        self.pe = torch.zeros(max_len, d_hid)
+
+        # Position order from `0` to `S - 1`.
+        position = torch.arange(0, max_len).unsqueeze(1)
+
+        # Compute the positional encodings once in log space.
+        # Shape : `(1, S, H)`
+        div_term = torch.exp(torch.arange(0, d_hid, 2) *
+                             -(math.log(10000.0) / d_hid))
+        self.pe[:, 0::2] = torch.sin(position * div_term)
+        self.pe[:, 1::2] = torch.cos(position * div_term)
+        self.pe = self.pe.unsqueeze(0)
+
+    def forward(self, src: torch.Tensor) -> torch.Tensor:
+        r"""Perform forward pass.
+
+        Parameters
+        ==========
+        src: torch.Tensor
+            Input sequence with shape ``(B, S, E)``.
+
+        Returns
+        =======
+        torch.Tensor
+            Same shape as input.
+
+        """
+        # Add positional encoding to each sequences.
+        # Input shape : `(B, S, H)`
+        # Output shape : `(B, S, H)`
+        pe = self.pe.detach().to(src.device)
+        output = src + pe[:, :src.size(1)]
+        return self.dropout(output)
+
+
+class TransformerModel(RNNModel):
```

Comment:
Since `Transformer` is not an `RNN` model, use `BaseModel` as base class instead.